### ISSUE TITLE
Which criterion is equivalent to softmax classifier

### ISSUE THREAD

**Issue by johnny5550822** [torch/nn#357]
Sorry, I am a bit confused, which criterion is equivalent to a softmax classifier?

**Comment by nicholas-leonard** [comment]
Output layer and criterion options (all are equivalent, 1 is most popular) :
 1. Linear + LogSoftMax + ClassNLLCriterion
 2. Linear + SoftMax + Log + ClassNLLCriterion
 3. Linear + CrossEntropyCriterion

**Comment by johnny5550822** [comment]
@nicholas-leonard  Let assume I am doing binary calssification, so the output of the first one is a probability distribution of two classes? Or log probability

**Comment by nicholas-leonard** [comment]
The output of LogSoftMax is a log probability. 

Note that in order to make effective use of SoftMax, you need to have at least two output units. If you output a single value between 0 and 1, then you can use : 

Linear + Sigmoid + BCECriterion.

**Comment by johnny5550822** [comment]
So that mean with #1, I have two outputs (probabilities), but they are in log form?

Can I do:
Linear + SoftMax + ClassNLLCriterion

**Comment by nicholas-leonard** [comment]
Yeah they are in log form.

No Linear + SoftMax + ClassNLLCriterion won't work.
If you absolutely want the output of your model to be a SoftMax, you can use [dpnn](https://github.com/nicholas-leonard/dpnn) to build a criterion that will do the log for you like this : 
```lua
criterion = nn.ModuleCriterion(nn.ClassNLLCriterion(), nn.Log())
```
So you would end up using this (which will work) : 
Linear + SoftMax + ModuleCriterion(ClassNLLCriterion, Log)
